### PR TITLE
Hmmmm. Bunch of changes. Copying plugin resources, few docs, api improvements etc

### PR DESCRIPTION
**Pull request by baev**

### CODE REVIEW

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy [hunk 1]
```diff
@@ -1,94 +1,162 @@
...
     public <T> void prepare(T object) {
-        if (!object) {
-            return
-        }
-        def plugins = storage.get(object.class, PreparePlugin)
-
-        plugins.each {
-            it.prepare(object)
-        }
+        preparePlugins.get(object?.class)*.prepare(object)
```
Comment (lanwen) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: what means star? (*)

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy [hunk 2]
```diff
@@ -1,94 +1,162 @@
...
 
-        Storage(PluginLoader loader, Injector injector) {
-            def plugins = loader.loadPlugins()
-            if (!plugins) {
-                return
+    /**
+     * Find all resources for plugin.
+     */
+    protected static List<URL> findPluginResources(ProcessPlugin plugin) {
+        def path = plugin.class.canonicalName.replaceAll('\\.', '/')
```
Comment (lanwen) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: why dont simply `replace('.', '/')`?

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy [hunk 3]
```diff
@@ -1,94 +1,162 @@
...
+        def pattern = ~"^$path/.+\$"
+        def result = []
+        for (def resource : ClassPath.from(plugin.class.classLoader).resources) {
+            if (resource.resourceName =~ pattern) {
```
Comment (lanwen) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: what means `=~`?

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy (hunk 1 above)
Comment (baev) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: group operation. Call method for each element

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy (hunk 2 above)
Comment (baev) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: =)

File: allure-report-data/src/test/java/ru/yandex/qatools/allure/data/AllureReportGeneratorTest.java [hunk 4]
```diff
@@ -112,15 +108,15 @@ public void shouldGenerateWithoutFailures() throws Exception {
...
+        File dataDirectory = new File(outputDirectory, ReportWriter.DATA_DIRECTORY_NAME);
         assertTrue("Data directory should be created", dataDirectory.exists());
 
-        assertThat(dataDirectory, contains(XUnitPlugin.XUNIT_JSON));
-        assertThat(dataDirectory, contains(TimelinePlugin.TIMELINE_JSON));
-        assertThat(dataDirectory, contains(BehaviorsPlugin.BEHAVIORS_JSON));
-        assertThat(dataDirectory, contains(DefectsPlugin.DEFECTS_JSON));
-        assertThat(dataDirectory, contains(EnvironmentPlugin.ENVIRONMENT_JSON));
-        assertThat(dataDirectory, contains(GraphPlugin.GRAPH_JSON));
+        assertThat(dataDirectory, contains("xunit.json"));
```
Comment (lanwen) [inline allure-report-data/src/test/java/ru/yandex/qatools/allure/data/AllureReportGeneratorTest.java]: why replace consts with strings?

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy (hunk 3 above)
Comment (baev) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: Creates matcher and calls `find()` method

File: allure-report-data/src/test/java/ru/yandex/qatools/allure/data/AllureReportGeneratorTest.java (hunk 4 above)
Comment (baev) [inline allure-report-data/src/test/java/ru/yandex/qatools/allure/data/AllureReportGeneratorTest.java]: I know u ask it =) I remove constants. Now plugin name specified in `@Plugin.Name` annotation, so constants is redundant

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy (hunk 2 above)
Comment (baev) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/PluginManager.groovy]: @lanwen see #561

**Pull request merged**